src/hooks/useLocalStorage.jsx

```diff
@@ -0,0 +1,33 @@
+import { useState } from "react";
+
+const useLocalStorage = (key, defaultValue) => {
+  const [localStorageValue, setLocalStorageValue] = useState(() => {
+    try {
+      const value = localStorage.getItem(key);
+      if (value) {
+        return JSON.parse(value);
+      } else {
+        localStorage.setItem(key, JSON.stringify(defaultValue));
+        return defaultValue;
+      }
+    } catch (error) {
+      localStorage.setItem(key, JSON.stringify(defaultValue));
+      return defaultValue;
+    }
+  });
+
+  const setLocalStorageStateValue = (valueOrFn) => {
+    let newValue;
+    if (typeof valueOrFn === "function") {
+      const fn = valueOrFn;
+      newValue = fn(localStorageValue);
+    } else {
+      newValue = valueOrFn;
+    }
+    localStorage.setItem(key, JSON.stringify(newValue));
+    setLocalStorageValue(newValue);
+  };
+  return [localStorageValue, setLocalStorageStateValue];
+};
+
+export default useLocalStorage;
```

Comment:
@NemishGorasiya refactor code like this.

`'use client';

import { useEffect, useState } from 'react';

const useLocalStorage = (key, initialValue) => {
  const storedValue = typeof window !== 'undefined' ? localStorage.getItem(key) : null;
  const initial = storedValue ? JSON.parse(storedValue) : initialValue;

  const [value, setValue] = useState(initial);

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(value));
  }, [key, value]);

  return [value, setValue];
};

export default useLocalStorage;
`